### ISSUE TITLE
Unit Tests to avoid regression in existing demo programs

### ISSUE THREAD

**Issue by sandraros** [abap2xlsx/abap2xlsx#894]
Proposing this issue as an open discussion.

I'd like to know your opinion about adding "unit" tests for the demo programs. That would imply:
- to store the XLSX files generated by the demo programs in the Web Repository (W3MI objects)
- to extract the code from the demonstration programs into new classes which can be called from the demo programs and from a test program
- to create a test program whose ABAP Unit test checks that the file generated by each demo program 

In the future, we could also use this feature to make sure that both "Writer" and "Huge Writer" generate the same result (currently we have many features which don't work in the Huge Writer), and the same for "Reader" and "Huge Reader". This 
[program](https://github.com/sapmentors/abap2xlsx/files/7577209/abap_program_to_compare_writer_and_huge_writer.txt) can be used as a starting point for discussion on Writer/Huge Writer, it runs 2 demos online and displays the diff, which works as explained in #891, and it also contains the test classes for them.

**Comment by AndreaBorgia-Abo** [comment]
I like the general idea and the few doubts I have will probably be swept aside once I see the test program.
Some questions I have:

- how fine-grained will the checks have to be? Some changes (moving an option within a xml node) are harmless, whereas other are potentially disruptive.
- how do we pick the reference files? Do we arbitrarily choose the currently generated versions or do you have more specific criteria to match?

**Comment by sandraros** [comment]
Thanks for looking into it. I have changed my mind about ABAP Unit, because what I propose is not about unit testing. So I switched to a simple program which is to be run on demand and displays the differences of actual versus expected.

I have created this draft #902 so that you can run the demo program `ZDEMO_EXCEL_CHECKER`. Note that you must install https://github.com/sandraros/zip-diff if you want to compare the ZIP at file level. If not, it just shows that there is a difference. By using a separate project, I can improve zip-diff any time, any fix will not pollute abap2xlsx.

1. Any change would be shown, except for fields which change at each generation, and for cells containing `sy-datum` and `sy-uzeit`.
2. I have created the reference files in `SMW0`, copying them from files generated by old versions of `ZDEMO_EXCEL##` programs, except those which fill the Excel file with current date and time, which are now forced to a fixed date and time when running through `ZDEMO_EXCEL_CHECKER`, so that the comparison can be done. These programs are: `ZDEMO_EXCEL_COMMENTS`, `ZDEMO_EXCEL1`, `ZTEST_EXCEL_IMAGE_HEADER`, `ZDEMO_EXCEL24`.

It's just a preview, all questions and comments welcome!

**Comment by AndreaBorgia-Abo** [comment]
The associated PR being a draft, I propose that I deal with it last, after the other PR in the queue have been reviewed. In the meantime, perhaps #891 could be closed and zip-diff included in the project documentation somewhere (Contributing?)

**Comment by sandraros** [comment]
Well, I kept the PR as draft because I think we are more in the process of discussion, and I did the PR so that to clarify my proposal. When we are clear about the goal and what the program should do exactly, we can complete the PR and pass it "final".

I would suggest to do #894 first because it will permit to make sure easily that there are no regressions with all new fixes.

#891 should not be closed, it's a different program. I proposed 2 different programs in #891 and #894, both use zip-diff, and both bring functionality which doesn't exist in zip-diff:
- #891: display screen with Excel workbook to the left and show differences in the XLSX file to the right after manual changes in Excel workbook and button pressed
- #894: show list of most abap2xlsx demo programs and indicate if there are regressions compared to MIME Objects stored in Web Repository
- zip-diff: display differences between 2 ZIP-like files (ZIP, XLSX, DOCX, etc.)

If possible, I'd prefer to update the Contributing part after we are clear about the program.

**Comment by AndreaBorgia-Abo** [comment]
> #891 should not be closed, it's a different program.

I ran zip-diff by itself once, saw that there was one option giving a display similar to the program snippet in #891 and thought that was all; my bad, I did not try the other option, now that I have used it with ZDEMO_EXCEL_CHECKER it all makes sense. 


> I proposed 2 different programs in #891 and #894, both use zip-diff

However the [program in 891](https://github.com/sapmentors/abap2xlsx/files/7612065/xlsx_diff_program_issue_891.txt) works even without zip-diff: did you mean that both display a diff or am I perhaps using an older version?


> I would suggest to do #894 first because it will permit to make sure easily that there are no regressions with all new fixes.

You mean #902 right? From a not-so-quick look at the code it looks like a good solution to the problem, without overburdening the rest of the code or the repository. I will start reviewing this in the coming days, then. Also, ok with updating the docs later on.

**Comment by sandraros** [comment]
Sorry, I created zip-diff while proposing #894, and I didn't rework #891 yet, to use zip-diff.

#902 if you prefer 😉. As I said, it's worth discussing first the issue (#894), the draft PR can be used to support the discussion.

Thanks! 👍

**Comment by AndreaBorgia-Abo** [comment]
Comments in no particular order:
* we'd probably need a small section in the contributing guidelines on the new way to write demo reports (interface and so on)
* I'm not sure I understand exactly why GET_NEXT_GENERATOR is empty except for ZDEMO_EXCEL15: once it gets there it runs through the rest of the rows in the table and the loop in the external program is bypassed
* I need to dig deeper into the workings of the generator class and maybe force spurious errors to see how it looks then
Will try again in the coming days.

**Comment by AndreaBorgia-Abo** [comment]
Moar comments:
* why the full class path for ty_zip_structure ? It's used only in the same classe, zcl_excel_generator.
* generator: if I am not mistaken, the loop in ZDEMO_EXCEL_CHECKER points to all the demos you can run as a user and the iteration done via GET_NEXT_GENERATOR serves to feed individual test cases for demos that might run with more than one file.
* forced an error and saw how it behaves, very nice.

All in all, I'd say the PR could be un-drafted.

**Comment by sandraros** [comment]
I do one answer for the whole points:

1. Documentation [created here](https://github.com/abap2xlsx/abap2xlsx/blob/sandraros/check-demo-regressions/docs/contributing/demo-programs.md) but not sure how to integrate it in a more general "coding" documentation which should wrap the variable naming standards and how demo programs work.
2. `GET_NEXT_GENERATOR` is a quick and dirty solution for `ZDEMO_EXCEL15` and for all future programs which generate more than one Excel file. Didn't think a lot about it, any suggestion welcome.
3. OK
4. `zcl_excel_demo_generator=>ty_zip_structure` is generated by ADT refactoring tool (maybe related to my old ABAP 7.52). I manually fixed it.
5. `ZDEMO_EXCEL_CHECKER` should test only demo programs which produce files of type Excel (not CSV), and only the ones which produce the same data all the time (so not `ZDEMO_EXCEL25`, and those which write the current time were modified to write a fixed time when ran by `ZDEMO_EXCEL_CHECKER`).
6. Thanks
7. The PR is un-drafted
Currently, I have kept one Excel file in the Web repository which is different from what is actually produced (`ZDEMO_EXCEL1`), just for the tests, we'll need to commit the latest Excel file.
